tools/clang_format_wrapper.py

```diff
@@ -0,0 +1,136 @@
+#!/usr/bin/env python
+"""Wrapper script for `git-clang-format` to avoid race conditions when
+merging/rebasing CPP files with merge conflicts."""
+from collections import namedtuple
+from os import environ as env
+from subprocess import check_output
+from typing import List, Iterable
+import argparse
+import enum
+import logging
+import sys
+
+logger = logging.getLogger(__name__)
+
+
+class GitState(enum.Enum):
+    """
+    Excerpt from https://git-scm.com/docs/git-status
+
+    For paths with merge conflicts, X and Y show the modification states
+    of each side of the merge. For paths that do not have merge conflicts,
+    X shows the status of the index, and Y shows the status of the work
+    tree. For untracked paths, XY are ??. Other status codes can be
+    interpreted as follows:
+    M = modified
+    A = added
+    D = deleted
+    R = renamed
+    C = copied
+    U = updated but unmerged
+    """
+
+    ADDED = "A"
+    COPIED = "C"
+    DELETED = "D"
+    MODIFIED = "M"
+    RENAMED = "R"
+    UNMODIFIED = " "
+    UNTRACKED = "?"
+    UPDATED_BUT_UNMERGED = "U"
+
+
+GitEdit = namedtuple("GitEdit", ["x", "y", "path"])
+
+
+def added_and_modified_paths() -> Iterable[str]:
+    """Inspect `git status` output.
+
+    Return paths to be modified by `git-clang-format`.
+    """
+
+    command = ["git", "status", "--porcelain", "--untracked-files=no"]
+    logger.debug("Executing `%s`", " ".join(command))
+    git_status_output = check_output(args=command, text=True)
+
+    edits = [
+        edit for edit in git_status_output.rstrip().split("\n") if edit != ""
+    ]
+    logger.debug(
+        "Found %d path(s) with state change in GIT status output", len(edits)
+    )
+
+    if edits:
+        logger.debug("Pending GIT edits:\n%s", "\n".join(edits))
+
+    return (
+        git_edit.path
+        for git_edit in (
+            GitEdit(edit[0:1], edit[1:2], edit[3:]) for edit in edits
+        )
+        if git_edit.x in (GitState.MODIFIED.value, GitState.ADDED.value)
+    )
+
+
+def run_clang_format(file_list: List[str]) -> bool:
+    """Execute git-clang-format for provided `file_list`.
+
+    Return True if `git-clang-format` has errors or has modified any file,
+    i.e. if it failed the formatting test.
+    """
+    if not file_list:
+        logger.info("No file to be formatted, exiting")
+        return
+
+    command = ["git-clang-format"]
+
+    src_commit = env.get("PRE_COMMIT_FROM_REF") or env.get("PRE_COMMIT_SOURCE")
+
+    if src_commit:
+        command.append(src_commit)
+
+    command.append("--")
+    command.extend(file_list)
+
+    logger.info("Executing `%s`", " ".join(command))
+    output = check_output(command, text=True)
+    logger.debug(output)
+
+    if output == "clang-format did not modify any files\n":
+        return False
+
+    return True
```

Comment:
```suggestion
    return output != "clang-format did not modify any files\n"
```

But that's just a minor code style issue. let's not hold up this PR longer.

Reply:
How about the "explicit > implicit" _zen_ suggestion? What you recommend looks like something you'd do in C rather than in Python to me.

Reply:
I see no difference in regards to explicity. The current version is more verbose though. I don't really care if we change it or not.

Reply:
More verbose is generally a sign of more explicit (even though I agree there's no linear correlation between the two concepts). In Python codebases that's usually what I see, i.e. write logic as you'd do in a more "natural" kind of language (i.e. English). Who cares if it takes away one two more lines of code? For how I see it, unless you want to win some kind of LoC contest, that should be irrelevant.

Reply:
Generally it's a good idea to reduce the line count as it makes stuff easier to maintain and has less room for possible bugs. In this case this just inverts the boolean logic of the original expression, so IMHO it makes sense to avoid this duplication. But as I said I don't really care, so we can leave it as is.

Btw, returning a boolean expression directly instead of wrapping it in an `if` block is not really uncommon in Python. For reference, some examples from the Python std library:
https://github.com/python/cpython/blob/578c3955e0222ec7b3146197467fbb0fcfae12fe/Lib/datetime.py#L44
https://github.com/python/cpython/blob/448325369ff73011d34d6c3a493014fe3ead8843/Lib/pathlib.py#L53
https://github.com/python/cpython/blob/8f1ed21ecf57cc8b8095d9d1058af2b9b3ed0413/Lib/weakref.py#L601
https://github.com/python/cpython/blob/8f1ed21ecf57cc8b8095d9d1058af2b9b3ed0413/Lib/weakref.py#L607
https://github.com/python/cpython/blob/8a64ceaf9856e7570cad6f5d628cce789834e019/Lib/glob.py#L206
https://github.com/python/cpython/blob/8a64ceaf9856e7570cad6f5d628cce789834e019/Lib/glob.py#L209
https://github.com/python/cpython/blob/8a64ceaf9856e7570cad6f5d628cce789834e019/Lib/glob.py#L213-L215

Reply:
A one liner might do magic but my argument is that usually it makes the codebase less maintainable, not more. Code should be made first of all for humans to be understood.

By the way, not a single example you linked from the stdlib makes use of the `!=` operator, which is exactly what I find less readable compared to the more verbose solution I propose.

Reply:
In the last patch I made it in oneliner which I think is also readable:

`return False if out == "clang-format did not modify any files\n" else True`

This way it should satisfy both the need for low LoC and readability

Reply:
I don't understand how `!=` is less readable than `==` to be honest. IMHO invert the output of the expression makes it a little bit harder to understand ("if the expression is true then the function returns false and the other way around" vs "the function returns expression result").

These all give the same result, but I personally find the last one the easiest to read: 
```python
False if out == "clang-format did not modify any files\n" else True
not (out == "clang-format did not modify any files\n")
out != "clang-format did not modify any files\n"
```

But again, this is not really important and we can keep it like that. It's still readable and understandable.

Reply:
The only example that reads well in English is the first one. I personally find `==` and `!=` quite different on a readability level and I'm pretty sure I'm not the only one (I believe that the brain to validate negation first searches for equality and than flips the outcome, so two mental steps instead of one).

Another example is that some other programming language like Ruby have the `unless` operator that many believe makes things easier to read (and write).

About how this bit is not important, I believe that not much of the discussion on this pull request was important at all on a functional perspective, the business logic has been almost untouched if I'm not wrong and the recommendations were about style and design. To me it has been an interesting dissertation on what's the meaning of _pythonic_ , what makes code easier to read (and maintain) and so on, and, most importantly, a way to understand what are the shared practices in Mixxx development.